### PR TITLE
Upload

### PR DESCRIPTION
**Pull request by leepc12**
Added the following features (4/25/2023)
- check status of object before uploading
- show status/error on both sheet and uploader sidebard
- prevent re-uploading of `released` accession
- reworked on how Global/sheet settings are stored: those settings can be shared by users and preserved even after making a copy
- passwords are stored in a secured per-user storage

Fixed
- relpath disappears when file is posted
- show number of matching files found (on the init button)
- hanging issue (it turns to be authorization problem, now script show errors when credentials are wrong/empty)

### CODE REVIEW

File: functions/Metadata.js [hunk 1]
```diff
@@ -23,10 +26,10 @@ function getTooltipForCommentedProp(prop) {
   }
 }
 
-function makeMetadataUrl(method, profileName, endpoint, identifyingVal) {
+function makeMetadataUrl(method, profileName, endpoint, identifyingVal, getSuffix="") {
```
Comment (khzlin) [inline functions/Metadata.js]: looks redundant

File: functions/Upload.js [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    if (json.hasOwnProperty(HEADER_COMMENTED_PROP_SKIP)) {
+      if (toBoolean(json[HEADER_COMMENTED_PROP_SKIP])) {
+        continue;
+      }
+    }
+    var [identifyingProp, identifyingVal, identifyingCol] =
+      findIdentifyingPropValColInRow(sheet, row, profile);
+
+    if (!identifyingProp || !identifyingVal) {
+      json[HEADER_COMMENTED_PROP_UPLOAD_STATUS] = "Missing identifying val.";
```
Comment (khzlin) [inline functions/Upload.js]: can you make this more user friendly? Not sure what `"Missing identifying val."` or `val` means if I didn't read the code.

**Review by khzlin — CHANGES_REQUESTED** [review]
Looks good. I left some comments for minor changes.

File: functions/Upload.js (hunk 2 above)
Comment (leepc12) [reply functions/Upload.js]: fixed

File: functions/Metadata.js (hunk 1 above)
Comment (leepc12) [reply functions/Metadata.js]: fixed

File: README.md [hunk 3]
```diff
@@ -104,3 +108,17 @@ Color and style represents a type of property.
 
 - <span style="text-decoration:underline">Underline</span>: Searchable property
 - ***Italic+Bold***: Array type property
+
+### Upload sidebar
```
Comment (khzlin) [inline README.md]: please change to add "File" in the title so users know what they are uploading. Maybe "File uploader" or "File Upload Sidebar"?

File: src/server/menu.js [hunk 4]
```diff
@@ -4,9 +4,10 @@ global.onOpen = () => {
 
   menu.addItem('Search', 'search');
   menu.addSeparator();
+  menu.addItem('Upload sidebar', 'uploadSidebar');
```
Comment (khzlin) [inline src/server/menu.js]: change to maybe "File uploader" or "File Upload Sidebar"?

**Review by khzlin — CHANGES_REQUESTED** [review]
One minor thing I would change is to rename "Upload Sidebar" to "File Upload Sidebar" so users know what they are actually submitting.

**Comment by khzlin** [comment]
found a few bugs:

1. relpath disappears when file is posted
2. file uploader side bar should show error message when there are no files
3. if there are no accession (even with an alias or uuid), file uploading is never complete. It just says “The file is uploading”.

File: src/server/menu.js (hunk 4 above)
Comment (leepc12) [reply src/server/menu.js]: fixed

File: README.md (hunk 3 above)
Comment (leepc12) [reply README.md]: fixed

**Review by khzlin — APPROVED** [review]

**Comment by khzlin** [comment]
Looks good to me!

**Pull request merged**